## IoT Botnet Detection

Due to Increasing usage of digital communication in this digital era, cyber security is crucial to maintain a high level of safety. To prevent an increasing number of cyber attacks, traditional security system firewalls, encryption is not enough. There is a need for intrusion detection systems that can integrate with traditional systems and assure a high level of security of data.

In [1]:
# Importing necssary modules
import pandas as pd
import numpy as np
import seaborn as sns
import os

In [2]:
# Data folder path
raw_data_path = '../rawdata'

The data contain 9 different IoT devices, each devices having benign / mallicious traffic data.

#### Device 1 - Danmini_Doorbell

Damini Doorbell data set contain 11 different csv file. each file represent different type of attack data. There is two malware used on this dataset. Mirai and Gafgyt.

* benign_traffic.csv - Benign Traffic
* mirai_attacks/ack.csv - Mallicious 1
* mirai_attacks/scan.csv - mallicious 2
* mirai_attacks/syn.csv - mallicious 3
* mirai_attacks/udp.csv - mallicious 4
* mirai_attacks/udpplain.csv - mallicious 5
* gafgyt_attacks/combo.csv - mallicious 6
* gafgyt_attacks/junk.csv - mallicious 7
* gafgyt_attacks/scan.csv - mallicious 8
* gafgyt_attacks/tcp.csv - mallicious 9
* gafgyt_attacks/udp.csv - mallicious 10

In [3]:
# Benign Traffic Data
damini_doorbell_path = raw_data_path + '/Danmini_Doorbell/'
damini_doorbell_benign = pd.read_csv(damini_doorbell_path + 'benign_traffic.csv')

In [4]:
damini_doorbell_benign.shape

(49548, 115)

In [5]:
damini_doorbell_benign.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49548 entries, 0 to 49547
Columns: 115 entries, MI_dir_L5_weight to HpHp_L0.01_pcc
dtypes: float64(115)
memory usage: 43.5 MB


In [6]:
damini_doorbell_benign.head()

MI_dir_L5_weight  MI_dir_L5_mean  MI_dir_L5_variance  MI_dir_L3_weight  \
0          1.000000       60.000000            0.000000          1.000000   
1          1.000000      354.000000            0.000000          1.000000   
2          1.857879      360.458980           35.789338          1.912127   
3          1.000000      337.000000            0.000000          1.000000   
4          1.680223      172.140917        18487.448750          1.793580   

   MI_dir_L3_mean  MI_dir_L3_variance  MI_dir_L1_weight  MI_dir_L1_mean  \
0       60.000000            0.000000          1.000000       60.000000   
1      354.000000            0.000000          1.000000      354.000000   
2      360.275733           35.923972          1.969807      360.091968   
3      337.000000            0.000000          1.000000      337.000000   
4      182.560279        18928.175300          1.925828      193.165753   

   MI_dir_L1_variance  MI_dir_L0.1_weight  ...  HpHp_L0.1_radius  \
0            0.000000            1.000000  ...          0.000000   
1            0.000000            1.000000  ...         34.095047   
2           35.991542            1.996939  ...        100.081513   
3            0.000000            1.000000  ...          0.000000   
4        19153.795810            1.992323  ...          0.000000   

   HpHp_L0.1_covariance  HpHp_L0.1_pcc  HpHp_L0.01_weight  HpHp_L0.01_mean  \
0                   0.0            0.0           1.000000        60.000000   
1                   0.0            0.0           5.319895       344.262695   
2                   0.0            0.0           6.318264       347.703087   
3                   0.0            0.0           1.000000       337.000000   
4                   0.0            0.0           1.000000        60.000000   

   HpHp_L0.01_std  HpHp_L0.01_magnitude  HpHp_L0.01_radius  \
0        0.000000             60.000000           0.000000   
1        4.710446            344.262695          22.188299   
2        9.034660            347.703087          81.625077   
3        0.000000            337.000000           0.000000   
4        0.000000             60.000000           0.000000   

   HpHp_L0.01_covariance  HpHp_L0.01_pcc  
0                    0.0             0.0  
1                    0.0             0.0  
2                    0.0             0.0  
3                    0.0             0.0  
4                    0.0             0.0  

[5 rows x 115 columns]

In [7]:
import glob

In [8]:
#Mirai_attacks data
damini_doorbell_mirai = pd.concat(map(pd.read_csv, glob.glob(damini_doorbell_path + 'mirai_attacks/*.csv')))

In [10]:
damini_doorbell_mirai.shape

(652100, 115)

In [11]:
damini_doorbell_mirai.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 652100 entries, 0 to 81981
Columns: 115 entries, MI_dir_L5_weight to HpHp_L0.01_pcc
dtypes: float64(115)
memory usage: 577.1 MB


In [12]:
damini_doorbell_mirai.head()

MI_dir_L5_weight  MI_dir_L5_mean  MI_dir_L5_variance  MI_dir_L3_weight  \
0          1.000000           566.0        0.000000e+00          1.000000   
1          1.996585           566.0        5.820766e-11          1.997950   
2          2.958989           566.0        0.000000e+00          2.975291   
3          3.958979           566.0        0.000000e+00          3.975285   
4          4.914189           566.0        1.164153e-10          4.948239   

   MI_dir_L3_mean  MI_dir_L3_variance  MI_dir_L1_weight  MI_dir_L1_mean  \
0           566.0        0.000000e+00          1.000000           566.0   
1           566.0        5.820766e-11          1.999316           566.0   
2           566.0        5.820766e-11          2.991729           566.0   
3           566.0        0.000000e+00          3.991727           566.0   
4           566.0        5.820766e-11          4.982654           566.0   

   MI_dir_L1_variance  MI_dir_L0.1_weight  ...  HpHp_L0.1_radius  \
0        0.000000e+00            1.000000  ...               0.0   
1        0.000000e+00            1.999932  ...               0.0   
2        5.820766e-11            2.999171  ...               0.0   
3        1.164153e-10            3.999171  ...               0.0   
4        5.820766e-11            4.998261  ...               0.0   

   HpHp_L0.1_covariance  HpHp_L0.1_pcc  HpHp_L0.01_weight  HpHp_L0.01_mean  \
0                   0.0            0.0                1.0            566.0   
1                   0.0            0.0                1.0            566.0   
2                   0.0            0.0                1.0            566.0   
3                   0.0            0.0                1.0            566.0   
4                   0.0            0.0                1.0            566.0   

   HpHp_L0.01_std  HpHp_L0.01_magnitude  HpHp_L0.01_radius  \
0             0.0                 566.0                0.0   
1             0.0                 566.0                0.0   
2             0.0                 566.0                0.0   
3             0.0                 566.0                0.0   
4             0.0                 566.0                0.0   

   HpHp_L0.01_covariance  HpHp_L0.01_pcc  
0                    0.0             0.0  
1                    0.0             0.0  
2                    0.0             0.0  
3                    0.0             0.0  
4                    0.0             0.0  

[5 rows x 115 columns]

In [9]:
# Gafgyt
damin_doorbell_gafgyt = pd.concat(map(pd.read_csv, glob.glob(damini_doorbell_path + 'gafgyt_attacks/*.csv')))

In [13]:
damin_doorbell_gafgyt.shape

(316650, 115)

In [14]:
damin_doorbell_gafgyt.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 316650 entries, 0 to 105873
Columns: 115 entries, MI_dir_L5_weight to HpHp_L0.01_pcc
dtypes: float64(115)
memory usage: 280.2 MB


In [15]:
damin_doorbell_gafgyt.head()

MI_dir_L5_weight  MI_dir_L5_mean  MI_dir_L5_variance  MI_dir_L3_weight  \
0          1.000000       98.000000        0.000000e+00          1.000000   
1          1.029000       98.000000        1.818989e-12          1.119520   
2          1.504156       76.725612        2.281808e+02          1.729662   
3          2.460087       75.617679        1.372200e+02          2.699075   
4          3.460055       75.150149        9.809937e+01          3.699054   

   MI_dir_L3_mean  MI_dir_L3_variance  MI_dir_L1_weight  MI_dir_L1_mean  \
0       98.000000            0.000000          1.000000       98.000000   
1       98.000000            0.000000          1.492583       98.000000   
2       79.499272          249.746357          2.294102       84.051188   
3       77.461807          164.269331          3.280499       80.987267   
4       76.525944          122.224798          4.280490       79.354915   

   MI_dir_L1_variance  MI_dir_L0.1_weight  ...  HpHp_L0.1_radius  \
0        0.000000e+00            1.000000  ...      0.000000e+00   
1        3.637979e-12            1.931640  ...      1.818989e-12   
2        2.517926e+02            2.904273  ...      0.000000e+00   
3        1.964467e+02            3.902546  ...      0.000000e+00   
4        1.592943e+02            4.902545  ...      0.000000e+00   

   HpHp_L0.1_covariance  HpHp_L0.1_pcc  HpHp_L0.01_weight  HpHp_L0.01_mean  \
0                   0.0            0.0           1.000000             98.0   
1                   0.0            0.0           1.992944             98.0   
2                   0.0            0.0           1.000000             66.0   
3                   0.0            0.0           1.000000             74.0   
4                   0.0            0.0           1.000000             74.0   

   HpHp_L0.01_std  HpHp_L0.01_magnitude  HpHp_L0.01_radius  \
0        0.000000             98.000000       0.000000e+00   
1        0.000001            138.592929       1.818989e-12   
2        0.000000            114.856432       0.000000e+00   
3        0.000000             74.000000       0.000000e+00   
4        0.000000             74.000000       0.000000e+00   

   HpHp_L0.01_covariance  HpHp_L0.01_pcc  
0                    0.0             0.0  
1                    0.0             0.0  
2                    0.0             0.0  
3                    0.0             0.0  
4                    0.0             0.0  

[5 rows x 115 columns]

There are 49548 rows of benign data, 652100 rows of mirai data, 316650 rows of gafgyt data. Mirai and gafgyt is combined dataframe of all 5 class of attacks.

In [18]:
damini_doorbell_benign.columns == damini_doorbell_mirai.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True])

In [19]:
damin_doorbell_gafgyt.columns == damini_doorbell_mirai.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True])

In [20]:
damini_doorbell_benign.columns == damin_doorbell_gafgyt.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True])

All three dataframe having same features